### ISSUE TITLE
Can't name a struct member with the name 't'

### ISSUE THREAD

**Issue by jritts** [dvidelabs/flatcc#82]
Adding a member variable with the name 't' appears to interfere with how flatcc generates the reader API.

For example given:
struct SVec3 {
   x	: float;
   y	: float;
   z	: float;
}
struct SAABB {
b  : SVec3;
t  : SVec3;
}

...then this declaration:
    __flatbuffers_define_struct_struct_field(SAABB, t, SVec3_struct_t)

generates a redefinition error in MSVC.

Is this a known problem?  Any way of working around it?

**Comment by mikkelfj** [comment]
I have not heard of it before, but given that it is MSVC, I'm surprised that I am not seeing more problems with it. (I have removed a ton of warnings though).

Running this with clang on MacOS does not result in any errors, so I assume that it is a problem with a MSVC header, or a problem with a definition in include/flatcc/portable specific to MSVC, but it is hard to tell without more information.

Adding a namespace to the schema might not work because `t` is a struct member, but you could try anyway. The accessor method would then be something like mynamespace_t(x), and if it is only the accessor method that is causing problems, it could work. Please let me know how it works out.

You can see how the macros are defined in the generated `flatbuffers_common_reader.h`.

**Comment by mikkelfj** [comment]
No, I spoke too early. It is clearly a problem in the flatcc generated header. I will have a look.

The macros necessarily have some local variables, and stupidly one of these are named t.

**Comment by mikkelfj** [comment]
There are many places where this needs to be fixed so I suggest you use longer names for now, and I'll try to come up with a proper fix later.

**Comment by jritts** [comment]
Thanks very much for taking a look so quickly - as you said, we'll use longer names for now and keep a look out for the fix.

**Comment by jritts** [comment]
A quick update, it appears a possibly related issue exists for a struct containing a variable named "start"

**Comment by mikkelfj** [comment]
thanks

Note that even with the tmp var name fixes, there will be potential name conflicts with operations.

I have tried to limit the surface, but it must be balanced against easy to use method names.
In the case of start, it is likely due to the start operation for starting a buffer of the given type. E.g. 
SAABB_start(...) starts building a SAABB buffer, but could also mean read a field named `start` . `create` is also likely to cause problems. Generally these are verbs while struct fields are often nouns. `scan` and `find` are also possible conficts.

In this case the use of start for buffer building probably wasn't very well thought ought, and perhaps it should be changed (a breaking change then). There are many other possible conflicts, for example you can't both have a struct member named `foo` and another named `foo_add`.

**Comment by mikkelfj** [comment]
Most names are suffixes to an existing field name The buffer start is different - so perhaps SAABB_buffer_start would be better. Suggestions are welcome.

**Comment by jritts** [comment]
Ok got it. 

Is it possible to tag all functions emitted by the generator, for example adding “get_” for all accessors?

We don’t require backwards compatibility, so assuming that’s a constraint for the library, do you have any pointers for how we might make such a change in our own fork?

Thanks again!

**Comment by mikkelfj** [comment]
It depends on your use case:

If you concern is with other code, you can modify namespaces - let me know if that is a concern - it is prepared for but not fully supported.

If it is a concern for a one-off collission with a name like `start` where you'd rather change the source than change the schema, it is easy: modify the generated `common` files, e.g. rename `start` to `do_start` or something. Then generate future headers without the flag that produces common files. You can also modify the src/compiler/codegen files, but that may be harder to maintain.

If you need to handle schema you cannot control at all, it is more difficult because there is no solution that will always work. Even if you add some prefix or suffix, names can conflict within that space too.

I don't think you need to be overly concerned with accessors - they would only conlict with a few other names suchs as `_add_*` and `_find_*`, `_vec_*`...
As I said `start` (and a few others) is a bit different because it does not follow a field name.

But, if `_get_` were to be added, it would a suffix like `SAABB_t_get(x)` not `SAABB_get_t(x)` because there is a certain grammar to it, both to easy code generators and to reduce conflict surface. Since an access operation can also be viewed as a cast operation, it is not unnatural to omit get here but it slightly increase the conflict potential.

Personally I think the getters are clumsy and are not worth the limited conflict potential (ignoring start that can be fixed), but if there is a general problem, we can add a _get suffix when moving to a major.

Now, say you want to make this `get` change anyway: it works like this:

The code generator asks a macro to define static inline methods - which are type safe. You have already identified the macro doing that for structs in your original post - the code generator generating the macro itself:

https://github.com/dvidelabs/flatcc/blob/de2da1ffca39d191aec160d91191ed2176a93b57/src/compiler/codegen_c_reader.c#L683-L686

The macro adds the schema namespace and the struct or table name as a prefix, then the field name, and finally any method name and then a variant name of that name to the static inline method.

Getters don't have method names added, but you can change that by changing the macro that defines the getters in the common_reader file, or in the src/compiler/ code generator the prints the common file (it prints it because it makes it possible to change name spaces).

If you want to contibute code where adding getters is a flatcc option, I can guide you to add the flags. The code generator then adds get based on that option.

Note that it is very tricky business to mess with these macros because of the layers of macro arguments. The struct accessor is the simplest possible thing, but there are also union vectors in tables etc.

**Comment by mikkelfj** [comment]
Maybe all conflicts can actually be removed by adding a get method? In that case the change might be worthwhile doing. At least as configurable option.
```
struct Bar {
    start: int;
    foo: int;
    foo_get: int;
    foo_add: int;
}
```

This would generate:
(simplified, there are many more - but it covers the grammar principles)

```
Bar_start(B)
Bar_end(B)
Bar_start_get(x)
Bar_start_add(B, x)
Bar_foo_get(x)
Bar_foo_add(B, x)
Bar_foo_get_get(x)
Bar_foo_get_add(B, x)
Bar_foo_add_get(B, x)
Bar_foo_add_add(B, x)
```

Wereas today you'd have:


```
Bar_start(B)
Bar_end(B)
Bar_start(x)
Bar_start_add(B, x)
Bar_foo(x)
Bar_foo_add(B, x)
Bar_foo_get(x)
Bar_foo_get_add(B, x)
Bar_foo_add(B, x)
Bar_foo_add_add(B, x)
```

**Comment by mikkelfj** [comment]
If such a change were made, it could be defined in

https://github.com/dvidelabs/flatcc/blob/master/config/config.h

```
#ifndef FLATCC_CGEN_ACCESSOR_METHOD
#define FLATCC_CGEN_ACCESSOR_METHOD _get
#endif
```

**Comment by mikkelfj** [comment]
So in the struct accessor for member structs it would look like:

```
#define AC FLATCC_CGEN_ACCESSOR
fprintf(out->fp, 
         "#define __%sdefine_struct_struct_field(N, NK, T)\\\n" 
         "static inline T N ## _ ## NK ## AC(N ## _struct_t t__tmp) { return t ? &(t__tmp->NK) : 0; }\n", 
         nsc); 
```

Note that I added the __tmp suffix for the original issue. But there are many other places to change.

**Comment by jritts** [comment]
Just saw these replies and wanted to first quickly say thank you for writing up so much information!!  I’ll dig into the code over the next few days and follow up again.

**Comment by mikkelfj** [comment]
Thanks, I analyze problems by writing, even if only to myself :-)

**Comment by mikkelfj** [comment]
You are gonna love this confict - which happens _after_ the tmp name issue was fixed (for this case at least):

```
$ ./x_reader.h:80:1: error: redefinition of 'SAABB_t' as different kind of symbol
__flatbuffers_define_struct_struct_field(SAABB, t, SVec3_struct_t)
^
./flatbuffers_common_reader.h:432:17: note: expanded from macro '__flatbuffers_define_struct_struct_field'
static inline T N ## _ ## NK(N ## _struct_t t__tmp__) { return t__tmp__ ? &(t__tmp__->NK) : 0; }
                ^
<scratch space>:54:1: note: expanded from here
SAABB_t
^
./x_reader.h:26:22: note: previous definition is here
typedef struct SAABB SAABB_t;fish: Unknown command './x_reader.h:80:1:'
```

So what is going on is that the `t` accessor function SAABB_t(x) now conflicts with the type `SAABB_t` where `_t` is the usual suffix to use when breaking a struct name out of the struct scope in C.

This, of course, will only happen for `t`, but I have no idea how to fix this, short of adding `_get`.

**Comment by mikkelfj** [comment]
Please have a go

https://github.com/dvidelabs/flatcc#potential-name-conflicts

https://github.com/dvidelabs/flatcc/commit/5c357274d0564d334ec8872d9c5281f876e68053

Note that I have not fixed potential name conflicts in the builders macros for `flatbuffers_common_builder.h`, but I think they are not very likely - and, I got tired of doing search replace. All the reader stuff should be good.

_get is not always generated, and the non-_get version can optionally be removed with -g, or hardcoded with FLATCC_CGEN_NO_CONFLICTS=1.

**Comment by jritts** [comment]
Wow that was fast -- thank you very much again.  It does fix the errors I was seeing (passing -g), and I'm waiting for my colleague to try it with his cases.

I made this change in portable/pstdalign.h to fix an error in MSVC:
~~~~
 /* https://lists.gnu.org/archive/html/bug-gnulib/2015-08/msg00003.html */
 #if defined(__cplusplus)
+#if !defined(_MSC_VER)
 #include <stdalign.h>
+#endif
 #if __cplusplus > 201103
~~~~

Also, we've integrated flatcc's cli tool into our ninja+gn based build system, and using the warning level at which we work, I made this change in portable/pstatic_assert.h:

~~~~
 #ifdef __COUNTER__
-#define static_assert(e, msg) enum { __PSTATIC_ASSERT_CONCAT(__COUNTER__, __LINE__) = 1/(!!(e)) }
+#define static_assert(e, msg) enum { __PSTATIC_ASSERT_CONCAT(__COUNTER__, __LINE__) = 1/((int)(!!(e))) }
 #else
 #include "pstatic_assert_scope.h"
~~~~

...in order to resolve this warning:
~~~~
warning 'warning C4804: '/': unsafe use of type 'bool' in operation'
~~~~

I'm not sure these are correct - if they are, I could create a pull request if you would like.

**Comment by mikkelfj** [comment]
Thanks - its easier / faster to fix while you still understand the problem.

It appears that you are using some version of MSVC in C++ mode (considering the boolean warning). I think your pstdalign patch at least needs a version check.

More generally, work is ongoing for adding a C++ test case, but it is currently failing on some of the things that you touch upon. If you are using C++, it would be a great help if you can try to apply your proposed changes to the PR https://github.com/dvidelabs/flatcc/pull/80 

I'm a bit puzzled that you apparently use C++ but do not have a built-in static_assert - maybe the fix should focus on using that instead?

My ci build for MSVC has to choose between MSVC 2010 and 2017, because appveyor won't do both. I just merged master into ci-more branch with extra builds - they all pass without warnings up to MSVC 2015 - but without the C++ test.

https://ci.appveyor.com/project/dvidelabs/flatcc/build/win%23698

**Comment by mikkelfj** [comment]
After some research, it appears no version of MSVC supports stdalign.h, so I committed the changes suggested - they were already under investigation due to as discussed in #80.

**Comment by mikkelfj** [comment]
Are you working with  _MSC_VER < 1600 (i.e. older than MSVC 2010)? Otherwise static_assert should be built in and then line with `1/!!(e)` should not be active. No attempts have been made to try to support MSVC < 1600.

**Comment by mikkelfj** [comment]
I'd still like to now why the static assert landed where it did, but I have now fixed this and other C++ issues and closed PR 80 with C++ test case which compiles cleanly from MSVC 2010 thorugh 2015 and clang and gcc on Linux.

**Comment by jritts** [comment]
>> It appears that you are using some version of MSVC in C++ mode 
Yes, this is the case.  I'm using MSVC 2017.

>> no version of MSVC supports stdalign.h, so I committed the changes 
>> have now fixed this and other C++ issues and closed PR 80
Neat, thank you!  I'll reset to tip code then.

>> I'd still like to now why the static assert landed where it did
Just got a chance to retest this now - sorry for my confusion, I must have made my change prior to this one, and I just didn't notice that it was no longer needed:
~~~~
commit cf0d593a62f3da383d6deb55b462172efd487e91
Author: Aidan Cully <github@aidan.users.panix.com>
Date:   Wed Jan 3 14:57:55 2018 -0500
    More robust static_assert (#69)
~~~~

I'll go ahead and close this ticket, then.  Thank you again for the help.

**Comment by mikkelfj** [comment]
Thanks for the update - glad it worked out.
BTW: you managed to put two major semi-sweedish companies into a single name SAAB and ABB.

**Comment by mikkelfj** [comment]
Note that I missed a few places that prevented the -g option from working with unions.
This has now been fixed on master in  dd5f97d in will be availble in the next release, as discussed in #96.
